sd: Enable eMMC Protocol [WIP]

decsny opened the pull request:
This is a work in progress PR to enable the eMMC protocol in zephyr. Tested on mimxrt595_evk_cm33 using the new tests/subsys/sd/mmc/ . The PR also does some small re-architecting of the SD subsystem. The main tasks left are to enable modern timings, add comments, and 4 bit bus option.

danieldegrasse committed:
sd: add stub for SDIO support

Add stub code for SDIO support, capable of verifying card responds to CMD5.
This commit also changes the architecture of the SDIO probe step to make
adding new protocol support more streamlined, and enable compiling out
support for undesired protocols.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
sd: split sdmmc common functions into sd_ops.c

split reusable portions of SDMMC protocol code into sd_ops.c, so other
SD protocols can use these functions directly without compiling in the
SDMMC subsystem.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

Commit:
sd: Changed KConfig Structure

Changed KConfig structure for SD:
- Better Menu Interface
- Changed symbol dependency structure

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
sd: Changed Init Header Structure

In sd subsystem, made one header for all the init functions
instead of having a bunch of header files with one function.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
boards: arm: mimxrt595_evk: Plumbs RT595 USDHC

- Adds the pin controls and ushdc settings in device tree
- Attaches clock to USDHC in soc.c
- Adds binding for mmc

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
boards: arm: Renamed NXP usdhc in imxrt5xx

The names of these peripherals in the device tree
did not match the Reference Manual for the RT500.

Also fixed a typo in a comment referring to USDHC which should have been
about USB.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
sd: MMC API, and MMC Spec

- Adds the declaration of the MMC API
- Adds MMC spec related declarations
- Adds some properties to sdhc dts binding for mmc

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
sd: Changes framework to support MMC

- Adds mmc.c
- Edits sd.c to init and probe MMC
- Adds mmc init to sd_init
- Some functions from sdmmc.c should be in sd_ops because
they can be used by both sdmmc and mmc.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
tests: sd: Adds MMC test and Kconfigs

Adds MMC test and Kconfigs for MMC
Note: MMC not yet implemented as of this commit

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

danieldegrasse commented on subsys/sd/mmc.c:
```diff
@@ -34,14 +50,491 @@ inline int mmc_ioctl(struct sd_card *card, uint8_t cmd, void *buf)
 int mmc_card_init(struct sd_card *card)
 {
 	int ret;
+	uint32_t ocr_arg = 0U;
```
Nit: you may want to check if the host controller uses SPI mode at this point, and return an error if it does (so SPI mode controllers cannot attempt to probe eMMC cards)

danieldegrasse commented on subsys/sd/mmc.c:
```diff
@@ -34,14 +50,491 @@ inline int mmc_ioctl(struct sd_card *card, uint8_t cmd, void *buf)
+			return ret;
+		}
+
+		//ret = mmc_bus_test_8_bit(card);
```
nit: should this be uncommented?

decsny replied:
actually it should be deleted because it doesn't exist

danieldegrasse commented on subsys/sd/mmc.c:
```diff
@@ -14,6 +14,22 @@
 #include "sd_utils.h"
 
+#define MMC_EXT_CSD_BYTES 512
+#define MMC_SWITCH_8_BIT_BUS_ARG (0xFC000000 & (0U << 26)) + \
```
May want to add a comment as to why these aren't in the spec file (assuming it is because they are specific command argument values)

danieldegrasse commented on subsys/sd/mmc.c:
```diff
@@ -14,6 +14,22 @@
 #include "sd_ops.h"
 #include "sd_utils.h"
 
+#define MMC_EXT_CSD_BYTES 512
```
May want to move this to the spec file, up to you

danieldegrasse commented on subsys/sd/mmc.c:
```diff
@@ -34,14 +50,491 @@ inline int mmc_ioctl(struct sd_card *card, uint8_t cmd, void *buf)
+			bus_test_data = bus_8bit_test_data;
+			bus_test_mask = bus_8bit_test_mask;
+			break;
+		case SDHC_BUS_WIDTH4BIT:
```
It seems you have support for testing 4 bit bus width, but not for setting it. Is that intentional, or just temporary?

decsny replied:
temporary

danieldegrasse reviewed (commented):
Overall very well done- just a few comments, primarily on ordering of commits. Core subsystem code looks well designed, as do the cleanups done to the SD stack

Aware, this is a draft, so feel free to wait to address these comments

decsny commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -315,6 +316,18 @@ zephyr_udc0: &usbhs {
+	pwr-gpios = <&gpio2 10 GPIO_ACTIVE_HIGH>;
+	cd-gpios = <&gpio2 9 GPIO_ACTIVE_LOW>;
+	mmc {
+		compatible = "zephyr,mmc-disk";
```
I know about this and have a few reasons to leave it like this, which I don't feel the need to list because I was already considering just squashing the commit after it onto this one

decsny commented on include/zephyr/sd/sd.h:
```diff
@@ -35,6 +35,7 @@ enum card_type {
 	CARD_SDMMC = 0, /*!< SD memory card */
 	CARD_SDIO = 1, /*!< SD I/O card */
 	CARD_COMBO = 2, /*!< SD memory and I/O card */
+	CARD_MMC = 3,
```
comments are something I still have to go back through and add in general

decsny commented on include/zephyr/sd/mmc.h:
```diff
@@ -0,0 +1,74 @@
+
+/**
+ * @file
+ * @brief Public API for SD memory card subsystem
```
yes

decsny commented on include/zephyr/sd/sd_spec.h:
```diff
@@ -538,6 +557,80 @@ struct sd_csd {
+ * MMC Timing Mode, encoded in EXT_CSD
+ */
+enum mmc_timing_mode {
+	MMC_BACK_TIMING = 0U,
```
why

decsny commented on subsys/sd/mmc.c:
```diff
@@ -0,0 +1,47 @@
+
+	ret = 0;
+
+	return ret;
```
yeah I had it like that originally I will change it back in next revision

stephanosio commented:
Please rebase to fix CI failure

decsny commented on tests/subsys/sd/mmc/src/main.c:
```diff
@@ -0,0 +1,248 @@
+	default:
+		zassert_unreachable("Card type is not known value");
+	}
+	/*
```
removed

decsny commented on tests/subsys/sd/mmc/testcase.yaml:
```diff
@@ -0,0 +1,13 @@
+    tags: sdhc
+    min_ram: 32
+    integration_platforms:
+      - mimxrt1064_evk
```
fixed next revision

Commit:
drivers: sdhc: Adds functionality to IMX USDHC

Adds some MMC support to IMX USDHC driver

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
sd: Enables eMMC in Zephyr

Enables eMMC protocol in Zephyr

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: sdhc: Fixes IMX SDHC DMA support

Enables DMA support for cacheless IMX systems.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
sd: Enables High Speed Timing for MMC protocol

- Enables HS Timing Mode for MMC protocol.
- Reads data about HS modes from EXT_CSD for MMC.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
tests: RT595 integration platform for MMC tests

Specifies RT595 as integration platform for MMC related tests

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
sd: zephyr.h is deprecated

Change from zephyr.h to kernel.h

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
updated at #50019
